### ISSUE TITLE
Commit Loss doesn't go down

### ISSUE THREAD

**Issue by ZekaiGalaxy** [lucidrains/meshgpt-pytorch#32]
Thanks for your work!

I tried to use the table dataset in ShapeNetV2 and test the code, the code works fine. And here is my issue: 

During training phase, if I test the code using small dataset, namely 100 objs, the commit loss can converge quickly.
But when I use the full table dataset, the commit loss gets higher and higher each epoch.
When I apply several augmentation strategies, things get worse.

I tried both Residual VQ and Residual LFQ, none of them work.

I guess it might be the autoencoder meet difficulties to quantize so many different faces. But I am not familiar with autoencoders or quantizations.

Anyone also meet this issue or are familiar with Residual VQs? Thanks

**Comment by lucidrains** [comment]
@ZekaiGalaxy congrats on replicating the results!

for the commit loss, i wouldn't worry about it so much. i've heard various reports of commit losses not going down (or in some cases even increasing), and yet the results look good

**Comment by lucidrains** [comment]
@ZekaiGalaxy for residual VQ, you can try turning off commit loss altogether. some papers i've read claimed that the EMA decay is all that is needed. again, only for VQ

**Comment by MarcusLoppe** [comment]
> @ZekaiGalaxy for residual VQ, you can try turning off commit loss altogether (set the commitment weight to `0.`). some papers i've read claimed that the EMA decay is all that is needed. again, only for VQ
@lucidrains 

I'm going to jump in here with probably irrelevant information, but I did some exploring and since the dim (in) for the LFQ is the same as the codebook_dims (out) the LFQ just uses the placeholder identity instead of a linear. Is this on purpose? 

class LFQ(Module):
```
has_projections = dim != codebook_dims
self.project_in = nn.Linear(dim, codebook_dims) if has_projections else nn.Identity()
self.project_out = nn.Linear(codebook_dims, dim) if has_projections else nn.Identity()
```

Complied model:
```
ResidualLFQ(
  (project_in): Linear(in_features=192, out_features=14, bias=True)
  (project_out): Linear(in_features=14, out_features=192, bias=True)
  (layers): ModuleList(
    (0-1): 2 x LFQ(
      (project_in): Identity()
      (project_out): Identity()
      (activation): Identity()
    )
  )
)
```

**Comment by lucidrains** [comment]
@MarcusLoppe yup that was just a design choice on my end, to automatically handle a different code dimension than model dimension

in this repository, i handle the projections outside of that package however

**Comment by MarcusLoppe** [comment]
> @MarcusLoppe yup that was just a design choice on my end, to automatically handle a different code dimension than model dimension
> 
> in this repository, i handle the projections outside of that package however

Oh I see the quantize is being called more than once.
But wouldn't that reduce the quantizer ability to learn? Since now it gets it's own residual output that itself produced meaning that it needs to learn two sorts of inputs; the actual input from encoder and the residual output from itself.

I'll read some more so i'm not talking out my ass

**Comment by lucidrains** [comment]
oh no it isn't, the projection happens external to the residual quantization

yea you should! it is [this seminal paper](https://arxiv.org/abs/2107.03312). assembly ai has a good writeup on it [too](https://www.assemblyai.com/blog/what-is-residual-vector-quantization/)

**Comment by MarcusLoppe** [comment]
> oh no it isn't, the projection happens external to the residual quantization
> 
> yea you should! it is [this seminal paper](https://arxiv.org/abs/2107.03312). assembly ai has a good writeup on it [too](https://www.assemblyai.com/blog/what-is-residual-vector-quantization/)

"Further Reading 
To experiment with RVQ in Pytorch, you can play around with the great open-source VQ repo by lucidrains."
Famous are we ? 🚀 

Okay last question since this feel like you are just teaching me and that is not the topic of this thread. 

Something feels wrong but I might just be looking in the wrong place.
The actual quantizer network seems just to be two linear;in and out. Then these are recursively used to project into the codebook.

_"In practice, a single quantizer layer would need a very large (indeed, exponentially large) codebook to represent the compressed data accurately. "_

According to the  assembly ai post, these two linear layers which have a total parameter count of 5k, can't really compress the data enough and will be very lossy, wouldn't adding more layers expand on the compression ability since the current network size might not have the ability to compress further after certain amount of data?


Your technique is probably better and you probably have read these, but this is what they used in the paper:

Autoregressive image generation using residual quantization
https://arxiv.org/pdf/2203.01941.pdf
Stacked Quantizers for Compositional Vector Compression
https://arxiv.org/pdf/1411.2173.pdf

**Comment by lucidrains** [comment]
@MarcusLoppe so they aren't actually referring to the linear layers, they are just building up an intuition why multiple quantization layers are better than one. quantization is parameter free and reliant on the encoder network before to do the compression

**Comment by lucidrains** [comment]
actually, vector quantization isn't parameter free, codebook is still learned. only the latest scalar quantization methods are parameter free
